evalml/pipelines/components/transformers/preprocessing/stl_decomposer.py

```diff
@@ -0,0 +1,340 @@
+
+    Args:
+        time_index (str): Specifies the name of the column in X that provides the datetime objects. Defaults to None.
+        degree (int): Not currently used.  STL 3x "degree-like" values.  None are able to be set at
+            this time. Defaults to 1.
+        seasonal_period (int): The number of entries in the time series data that corresponds to one period of a
+            cyclic signal.  For instance, if data is known to possess a weekly seasonal signal, and if the data
+            is daily data, seasonal_period should be 7.  For daily data with a yearly seasonal signal, seasonal_period
+            should be 365.  For compatibility with the underlying STL algorithm, must be odd. If an even number
+            is provided, the next, highest odd number will be used. Defaults to 7.
+        random_seed (int): Seed for the random number generator. Defaults to 0.
+    """
+
+    name = "STL Decomposer"
+    hyperparameter_ranges = {}
+
+    modifies_features = False
+    modifies_target = True
+
+    def __init__(
+        self,
+        time_index: str = None,
+        degree: int = 1,  # Currently unused.
+        seasonal_period: int = 7,
+        random_seed: int = 0,
+        **kwargs,
+    ):
+        self.logger = logging.getLogger(__name__)
+
+        # Programmatically adjust seasonal_period to fit underlying STL requirements,
+        # that seasonal_period must be odd.
+        if seasonal_period % 2 == 0:
+            self.logger.warning(
+                f"STLDecomposer provided with an even period of {seasonal_period}"
+                f"Changing seasonal period to {seasonal_period+1}",
+            )
+            seasonal_period += 1
+
+        super().__init__(
+            component_obj=None,
+            random_seed=random_seed,
+            degree=degree,
+            seasonal_period=seasonal_period,
+            time_index=time_index,
+            **kwargs,
+        )
+
+    def _check_oos_past(self, y):
+        """Function to check whether provided target data is out-of-sample and in the past."""
+        if y.index[-1] < self.trend.index[0]:
+            raise ValueError(
+                f"STLDecomposer cannot transform/inverse transform data out of sample and before the data used"
+                f"to fit the decomposer."
+                f"\nRequested date range: {str(y.index[0])}:{str(y.index[-1])}."
+                f"\nSample date range: {str(self.trend.index[0])}:{str(self.trend.index[-1])}.",
+            )
+
+    def _project_trend(self, y):
+        """Function to project the in-sample trend into the future."""
+        self._check_oos_past(y)
+
+        # Determine how many units forward to project by finding the difference,
+        # in index values, between the requested target and the fit data.
+        datetime_difference = y.index[-1] - self.trend.index[-1]
+        datetime_delta = pd.to_timedelta(1, self.frequency)
+        units_forward = int(datetime_difference / datetime_delta)
```

Comment:
```suggestion
        units_forward = len(pd.date_range(start=self.trend.index[-1], end=y.index[-1], freq=self.frequency))
```
`pd.to_timedelta` does not handle all frequencies - most notably nothing longer than a week. If you run this with monthly data (month start, to be exact), the `MS` frequency gets interpreted as milliseconds, causing errors.